internal/promotion/runner/builtin/tar_extractor.go

```diff
@@ -0,0 +1,336 @@
+	stripComponents := int64(0)
+	if cfg.StripComponents != nil {
+		stripComponents = *cfg.StripComponents
+		if stripComponents > 0 {
+			logger.Debug("stripping path components", "count", stripComponents)
+		}
+	}
+
+	// Track total size extracted
+	var totalExtractedSize int64 = 0
+
+	for {
+		header, err := tarReader.Next()
+		if err == io.EOF {
+			break // End of archive
+		}
+		if err != nil {
+			return promotion.StepResult{Status: kargoapi.PromotionStepStatusErrored},
+				fmt.Errorf("error reading tar: %w", err)
+		}
+
+		// Skip if this is not a regular file, symlink, or directory
+		if header.Typeflag != tar.TypeReg && header.Typeflag != tar.TypeDir && header.Typeflag != tar.TypeSymlink {
+			logger.Debug("skipping non-regular file", "path", header.Name, "type", header.Typeflag)
+			continue
+		}
+
+		// Handle stripping components if specified
+		targetName := header.Name
+		if stripComponents > 0 {
+			// Get parts of the path
+			parts := strings.Split(header.Name, "/")
+			// Only strip if we have enough components
+			if len(parts) > int(stripComponents) {
+				targetName = strings.Join(parts[stripComponents:], "/")
+			} else {
+				// Skip this file if we don't have enough components
+				logger.Debug("skipping file with insufficient path components", "path", header.Name)
+				continue
+			}
+		}
+
+		// Skip any empty targetName which can happen if we're processing a directory entry
+		if targetName == "" || targetName == "/" {
+			continue
+		}
+
+		// Validate the target path to prevent directory traversal
+		if !t.validRelPath(targetName) {
+			logger.Debug("skipping file with unsafe path", "path", targetName)
+			continue
+		}
+
+		// Simple check for exact filename match for ignore patterns
+		if cfg.Ignore != "" && filepath.Base(targetName) == strings.TrimSpace(cfg.Ignore) {
+			logger.Debug("ignoring exact match path", "path", targetName)
+			continue
+		}
+
+		// Check more complex patterns using gitignore matcher
+		isDir := header.Typeflag == tar.TypeDir
+		pathParts := strings.Split(targetName, "/")
+		if matcher.Match(pathParts, isDir) {
+			logger.Debug("ignoring path based on pattern", "path", targetName)
+			continue
+		}
+
+		// Create destination directory for files and links
+		targetPath := filepath.Join(outPath, targetName)
+
+		// Double check the target path is within the output directory
+		relPath, err := filepath.Rel(outPath, targetPath)
+		if err != nil || strings.HasPrefix(relPath, "..") || strings.HasPrefix(relPath, "/") {
+			logger.Debug("skipping path escaping target directory", "path", targetName)
+			continue
+		}
+
+		destDir := filepath.Dir(targetPath)
+		if err := os.MkdirAll(destDir, 0755); err != nil {
+			return promotion.StepResult{Status: kargoapi.PromotionStepStatusErrored},
+				fmt.Errorf("failed to create directory %s: %w", destDir, err)
+		}
+
+		switch header.Typeflag {
+		case tar.TypeDir:
+			// Create directory
+			if err := os.MkdirAll(targetPath, 0755); err != nil {
```

Comment:
`0o755` can probably be moved to a `const` (and then also be used above). (See also comment below about creating files). I also think we might want to use `0o750` instead.